[DDP-596][Self-service error codes] Add migration guide (#11632)
digital-asset/daml: docs/source/error-codes/self-service/index.rst

   :hidden:

.. list-all-error-codes::

Self-Service Error Codes Migration Guide
########################################

> Using self-service error codes means that some of the gRPC error codes returned from service methods change as well.

Reviewer: ```suggestion
The introduction of the self-service error codes means that some of the gRPC error codes returned from service methods change.
```

Consult the table below for details on those changes.

> Note that the table below contains entries only for cases when a gRPC codes was changed.

Reviewer: ```suggestion
Note that the table below contains entries only for cases where a gRPC codes was changed.
```

> For example, if a service method was returning gRPC error code ``CODE_A`` for some error
> conditions which are now signalled by ``CODE_B``, it could be the case that there are another error conditions for which ``CODE_A`` is still being returned,
> and only an entry for the change from ``CODE_A`` in ``CODE_B`` is included in the table below.

Reviewer: ```suggestion
For example, a service method previously returning gRPC error code ``CODE_A`` may now return 
``CODE_B`` for some error condition while returning ``CODE_A`` for other conditions,
and only an entry for the change from ``CODE_A`` in ``CODE_B`` is included in the table below.
```

|PackageManagementService.uploadDarFile         |ABORTED                         |DEADLINE_EXCEEDED                     |DEADLINE_EXCEEDED can now be returned when a time out was reached.                                                       |REQUEST_TIME_OUT                             |
+-----------------------------------------------+--------------------------------+--------------------------------------+-------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+
|PackageManagementService.uploadDarFile         |INVALID_ARGUMENT                |FAILED_PRECONDITION                   |FAILED_PRECONDITION can now be returned when a package upload was rejected.                                              |PACKAGE_UPLOAD_REJECTED                      |
+-----------------------------------------------+--------------------------------+--------------------------------------+-------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+

> |ParticipantPruningService.prune                |INVALID_ARGUMENT                |OUT_OF_RANGE                          |OUT_OF_RANGE can now be returned when the supplied offset was out of range.                                              |REQUESTED_OFFSET_OUT_OF_RANGE                |

Reviewer: ```suggestion
|ParticipantPruningService.prune                |INVALID_ARGUMENT                |OUT_OF_RANGE                          |OUT_OF_RANGE can now be returned when the supplied pruning offset was out of range.                                              |REQUESTED_OFFSET_OUT_OF_RANGE                |
```

|PartyManagementService.allocateParty           |ABORTED                         |DEADLINE_EXCEEDED                     |DEADLINE_EXCEEDED can now be returned when a time out was reached.                                                       |REQUEST_TIME_OUT                             |
+-----------------------------------------------+--------------------------------+--------------------------------------+-------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+
|SubmissionService.submit                       |ABORTED                         |ALREADY_EXISTS                        |ALREADY_EXISTS can now be returned when there was a duplicate contract key during interpretation.                        |DUPLICATE_CONTRACT_KEY_DURING_INTERPRETATION |
+-----------------------------------------------+--------------------------------+--------------------------------------+-------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+

> |SubmissionService.submit                       |ABORTED                         |INTERNAL                              |INTERNAL can now be returned when validation fails on replay mismatch.                                                   |LEDGER_API_INTERNAL_ERROR                    |

Reviewer: ```suggestion
|SubmissionService.submit                       |ABORTED                         |INTERNAL                              |INTERNAL can now be returned when validation fails on a mismatch during relay of the submitted transaction.                                                   |LEDGER_API_INTERNAL_ERROR                    |
```
